langchain/llms/sagemaker_async_endpoint.py

```diff
@@ -0,0 +1,176 @@
+
+        super().__init__(**kwargs)
+        region = self.region_name
+        account = boto3.client("sts").get_caller_identity()["Account"]
+        self.input_bucket = f"sagemaker-{region}-{account}" if input_bucket == "" else input_bucket
+        self.input_prefix = f"async-endpoint-outputs/{self.endpoint_name}" if input_prefix == "" else input_prefix
+        self.max_request_timeout = max_request_timeout
+        self.s3_client = boto3.client("s3")
+        self.sm_client = boto3.client("sagemaker")
+
+    def _call(
+        self,
+        prompt: str,
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any
+    ) -> str:
+        """
+        Call out to Sagemaker asynchronous inference endpoint.
+        Args:
+            prompt: The prompt to use for the inference.
+            stop: The stop tokens to use for the inference.
+            run_manager: The run manager to use for the inference.
+            kwargs: Keyword arguments to pass to the SagemakerEndpoint class.
+        Returns:
+            The output from the Sagemaker asynchronous inference endpoint.
+        """
+        # Parse the SagemakerEndpoint class arguments
+        _model_kwargs = self.model_kwargs or {}
+        _model_kwargs = {**_model_kwargs, **kwargs}
+        _endpoint_kwargs = self.endpoint_kwargs or {}
+
+        # Transform the input to match SageMaker expectations
+        body = self.content_handler.transform_input(prompt, _model_kwargs)
+        content_type = self.content_handler.content_type
+        accepts = self.content_handler.accepts
+
+        # Verify if the endpoint is running
+        response = self.sm_client.describe_endpoint(EndpointName=self.endpoint_name)
+        endpoint_is_running = response["ProductionVariants"][0]["CurrentInstanceCount"] > 0
+
+        # If the endpoint is not running, send an empty request to "wake up" the endpoint
+        test_data = b""
+        test_key = os.path.join(self.input_prefix, "test")
+        self.s3_client.put_object(Body=test_data, Bucket=self.input_bucket, Key=test_key)
+        if not endpoint_is_running:
+            self.client.invoke_endpoint_async(
+                EndpointName=self.endpoint_name,
+                InputLocation=f"s3://{self.input_bucket}/{test_key}",
+                ContentType=content_type,
+                Accept=accepts,
+                InvocationTimeoutSeconds=self.max_request_timeout,  # timeout of 60 seconds to detect if it's not running yet
+                **_endpoint_kwargs,
+            )
+            raise Exception("The endpoint is not running. Please check back in approximately 10 minutes.")
+        else:
+            print("Endpoint is running! Proceeding to inference.")
+
+        # Send request to the async endpoint
+        request_key = os.path.join(self.input_prefix, f"request-{str(uuid.uuid4())}")
```

Comment:
This might not be safe from collision in a multi-user scenario; should include a timestamp component.